### DISCUSSION TITLE
Sync compute shader for reading data from buffer.

### DISCUSSION THREAD

**Discussion by Anshi701** [vsg-dev/VulkanSceneGraph#1130] [Q&A]
Hello! I have a command graph with [queueFlags = VK_QUEUE_GRAPHICS_BIT | VK_QUEUE_COMPUTE_BIT]. In the graph there is some graphics draw nodes, and after that, node with compute shader. In the Comp Shader i am filling some storage buffer to make atomic counter what is created like that: 
``` C++
auto bufferSizeAtom = 4;
auto bufferAtom = vsg::createBufferAndMemory(device, bufferSizeAtom, VK_BUFFER_USAGE_STORAGE_BUFFER_BIT | VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT, VK_SHARING_MODE_EXCLUSIVE, VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
auto bufferMemoryAtom = bufferAtom->getDeviceMemory(device->deviceID);
```
After record and present stages i call function to read data from that buffer like that: 
``` C++
auto atom = vsg::MappedData<vsg::uintArray>::create(bufferMemoryAtom, 0, 0, vsg::Data::Properties{ VK_FORMAT_R32_UINT }, bufferSizeAtom); 
uint cnter2 = atom->at(0);
```
And there is a problem:
 - cnter2 value is varying with frames, and is not correct one.
 - if i put sleep with 10 ms before mapping, value is correct
If i understand that right, compute shader is simply not done yet to the moment after record() and present() functions.
What do i need to do to sync computations and read?

**Comment by robertosfield** [comment]
I'm afraid your description is too open ended for us to guess what you are trying to achieve, what your have implemented and what is happening on your system.

The best I can suggest is to read up on [Vulkan synchronization](https://www.khronos.org/blog/understanding-vulkan-synchronization), it's a bit of complex topic so be prepared to spend the time to understand it, even now I'm not an expert.  Most VSG usage doesn't require intimate knowledge of the synchronization side just occasionally you'll need to dig into the details.   

Broadly speaking Vulkan has:
1. vk/vsg::Semaphore for command buffers that don't require CPU intervention, 
2. vk/vsg::Fence for synchronizing the CPU with GPU
3. vk/vsg::Events that can be used for both CPU and inter GPU/command buffer synchronization
4. vk/vsg::PipelineBarrier that can be used for memory/image/buffer barriers.

With complex topics it's hard to help folks as there are so many unknowns and seeing the problem first hand with code and data that reproduces is often the difference between others being able to understand enough to help and not being able to help at all. It may be that you are attempting to do something that others haven't tried before, so this makes it more critical that others will find it helpful to have an simplified example that illustrates what you are doing.

Think about it in terms - you have 100% of the information about what you are trying to do, how you are doing and how it's failing, while others might have 1% of an idea of what you are doing so asking for help from folks that only have that 1% understand isn't likely to end in much help for anyone.

**Reply by Anshi701** [reply]
I understand what it is hard to help with little information. Let me then ask that question: "Is viewer->recordAndSubmit() function guarantees what all gpu commands are done (graphics and compute)?"

**Reply by robertosfield** [reply]
No. 

Vulkan/VSG allows you to do lots of different operations in different ways.  

Where possible the the Viewer class will attempt to synchronization the standard scene graph usage, but this is not the same as guaranteeing that all possible operations that you might set up are going to my synchronized correctly.  The RecordAndSubmitTask object allows you to specify some aspect of synchronization. 

Again this is big topic, there is lots of different ways of doing things, it's very open ended.  The VSG will try to help you where possible but it can't hide all the complexity all the time.

**Reply by Anshi701** [reply]
I get it, thanks for your answers and time!

**Reply by robertosfield** [reply]
The most helpful thing would be to create an example for vsgExamples that illustrates your usage case, then others can see it first hand and perhaps try to get it work. Once it's working both your and the rest of the community will benefit from having this particular usage case fleshed out.
